sips/sip-030/sip-030-wallet-interface.md

```diff
@@ -0,0 +1,532 @@
+  value: object // Clarity value
+}
+```
+
+`0x08` `err` (aka "response err")
+
+```ts
+{
+  type: 'err',
+  value: object // Clarity value
+}
+```
+
+`0x09` `none` (aka "optional none")
+
+```ts
+{
+  type: 'none',
+}
+```
+
+`0x0a` `some` (aka "optional some")
+
+```ts
+{
+  type: 'some',
+  value: object // Clarity value
+}
+```
+
+`0x0b` `list`
+
+```ts
+{
+  type: 'list',
+  value: object[] // Array of Clarity values
+}
+```
+
+`0x0c` `tuple`
+
+```ts
+{
+  type: 'tuple',
+  value: Record<string, object> // Record of Clarity values
+}
+```
+
+`0x0d` `ascii`
+
+```ts
+{
+  type: 'ascii',
+  value: string // ASCII-compatible string
+}
+```
+
+`0x0e` `utf8`
+
+```ts
+{
+  type: 'utf8',
+  value: string
+}
+```
+
+#### Post-conditions
+
+`0x00` STX
+
+```ts
+{
+  type: 'stx-postcondition',
+  address: 'origin' | string | `${string}.${string}`, // Stacks c32-encoded, with optional contract name suffix
+  condition: 'eq' | 'gt' | 'gte' | 'lt' | 'lte',
+  amount: string // `bigint` compatible, amount in micro-STX
+}
+```
+
+`0x01` Fungible token
+
+```ts
+{
+  type: 'ft-postcondition',
+  address: 'origin' | string | `${string}.${string}`, // Stacks c32-encoded, with optional contract name suffix
+  condition: 'eq' | 'gt' | 'gte' | 'lt' | 'lte',
+  asset: `${string}.${string}::${string}` // Stacks c32-encoded address, with contract name suffix, with asset suffix
+  amount: string // `bigint` compatible, amount in lowest integer denomination of fungible token
+}
+```
+
+`0x02` Non-fungible token
+
+```ts
+{
+  type: 'nft-postcondition',
+  address: 'origin' | string | `${string}.${string}`, // Stacks c32-encoded, with optional contract name suffix
+  condition: 'sent' | 'not-sent',
+  asset: `${string}.${string}::${string}` // address with contract name suffix with asset suffix, Stacks c32-encoded
+  assetId: object, // Clarity value
+}
+```
+
+#### Test vectors
+
+Listed below are some examples of the potentially unclear representations:
+
+- `u12` = `{ type: "uint", value: "12" }`
+- `0xbeaf` = `{ type: "ascii", value: "hello there" }`
+- `"hello there"` = `{ type: "ascii", value: "hello there" }`
+- `(list 4 8)` =
+  ```
+  {
+    type: "list",
+    value: [
+      { type: "int", value: "4"},
+      { type: "int", value: "8"},
+    ]
+  }
+  ```
+- `(err u4)` =
+  ```
+  {
+    type: "err",
+    value: { type: "uint", value: "4"},
+  }
+  ```
+- "sends more than 10000 uSTX" =
+  ```
+  {
+    type: "stx-postcondition",
+    address: "STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6",
+    amount: "10000",
+    condition: "gt"
+  }
+  ```
+- "does not send the `12` TKN non-fungible token" =
+  ```
+  {
+    type: "ntf-postcondition",
+    address: "STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6.vault"
+    asset: "STB44HYPYAT2BB2QE513NSP81HTMYWBJP02HPGK6.tokencoin::tkn",
+    assetId: { type: "uint", value: "12" }
+    condition: "not-sent"
+  }
+  ```
+
+### Provider registration
+
+Wallets can register their aliased provider objects however they see fit.
```

Comment:
Is this specific enough? Currently, different wallets use different methods and are not working nicely together.

Reply:
Yeah, I agree we should add more on provider discovery.